File: web/server/test-secp-256.py
from cryptography.hazmat.primitives.asymmetric import ec

# === P-256 curve parameters ===
P = 0xffffffff00000001000000000000000000000000ffffffffffffffffffffffff
A = 0xffffffff00000001000000000000000000000000fffffffffffffffffffffffc
B = 0x5ac635d8aa3a93e7b3ebbd55769886bc651d06b0cc53b0f63bce3c3e27d2604b
N = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
Gx = 0x6b17d1f2e12c4247f8bce6e563a440f277037d812deb33a0f4a13945d898c296
Gy = 0x4fe342e2fe1a7f9b8ee7eb4a7c0f9e162bce33576b315ececbb6406837bf51f5


# === Pure-Python elliptic curve operations over P-256 ===

INF = None  # point at infinity representation


def inv_mod(k: int, p: int) -> int:
	"""Modular inverse using extended Euclidean algorithm."""
	if k == 0:
		raise ZeroDivisionError("division by zero")
	if k < 0:
		return p - inv_mod(-k, p)

	s, old_s = 0, 1
	t, old_t = 1, 0
	r, old_r = p, k

	while r != 0:
		q = old_r // r
		old_r, r = r, old_r - q * r
		old_s, s = s, old_s - q * s
		old_t, t = t, old_t - q * t

	# old_r is gcd(k, p); for a prime field and nonzero k, this must be 1
	if old_r != 1:
		raise ValueError("inverse does not exist")

	return old_s % p


def point_double(point):
	"""Point doubling on the curve y^2 = x^3 + A*x + B over field P."""
	if point is INF:
		return INF

	x1, y1 = point
	if y1 % P == 0:
		return INF

	m = ((3 * x1 * x1 + A) * inv_mod(2 * y1, P)) % P
	x3 = (m * m - 2 * x1) % P
	y3 = (m * (x1 - x3) - y1) % P
	return x3, y3


def point_add(p1, p2):
	"""Point addition on the curve."""
	if p1 is INF:
		return p2
	if p2 is INF:
		return p1

	x1, y1 = p1
	x2, y2 = p2

	# P + (-P) = INF
	if x1 == x2 and (y1 + y2) % P == 0:
		return INF

	if x1 == x2 and y1 == y2:
		return point_double(p1)

	m = ((y2 - y1) * inv_mod((x2 - x1) % P, P)) % P
	x3 = (m * m - x1 - x2) % P
	y3 = (m * (x1 - x3) - y1) % P
	return x3, y3


def scalar_mult(k: int, point):
	"""Multiply a point by an integer k using double-and-add."""
	if point is INF:
		return INF

	k = k % N
	if k == 0:
		return INF

	result = INF
	addend = point

	while k:
		if k & 1:
			result = point_add(result, addend)
		addend = point_double(addend)
		k >>= 1

	return result


def is_on_curve(point) -> bool:
	if point is INF:
		return True
	x, y = point
	return (y * y - (x * x * x + A * x + B)) % P == 0


# === Use cryptography only to generate a random key, then verify with pure Python math ===

# Generate new EC key pair on secp256r1 (aka P-256)
private_key = ec.generate_private_key(ec.SECP256R1())

nums = private_key.private_numbers()
pub_nums = nums.public_numbers

d = nums.private_value      # private scalar as int
x = pub_nums.x              # public x-coordinate as int
y = pub_nums.y              # public y-coordinate as int

print("d =", d)
print("x (cryptography) =", x)
print("y (cryptography) =", y)

G = (Gx, Gy)
assert is_on_curve(G), "Base point G is not on curve"

calc_x, calc_y = scalar_mult(d, G)

print("x (pure Python) =", calc_x)
print("y (pure Python) =", calc_y)
print("x matches:", calc_x == x)
print("y matches:", calc_y == y)

File: web/server/test-secp-256.js
// P-256 (secp256r1) curve parameters as BigInt

const P = 0xffffffff00000001000000000000000000000000ffffffffffffffffffffffffn;
const A = 0xffffffff00000001000000000000000000000000fffffffffffffffffffffffcn; // -3 mod P
const B = 0x5ac635d8aa3a93e7b3ebbd55769886bc651d06b0cc53b0f63bce3c3e27d2604bn;
const N = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551n;
const G = [
    0x6b17d1f2e12c4247f8bce6e563a440f277037d812deb33a0f4a13945d898c296n,
    0x4fe342e2fe1a7f9b8ee7eb4a7c0f9e162bce33576b315ececbb6406837bf51f5n
];

// Point at infinity representation
const INF = null; // null means point at infinity

function hexToBytes(hex) {
    if (hex.length % 2 !== 0) {
        throw new Error("Invalid hex string");
    }
    const bytes = new Uint8Array(hex.length / 2);
    for (let i = 0; i < hex.length; i += 2) {
        bytes[i / 2] = parseInt(hex.substr(i, 2), 16);
    }
    return bytes;
}

function bytesToHex(bytes) {
    return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
}

function bytesToBase64(bytes) {
    let binary = '';
    for (let i = 0; i < bytes.byteLength; i++) {
        binary += String.fromCharCode(bytes[i]);
    }
    return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

function base64ToBytes(base64) {
    const binary = atob(base64.replace(/-/g, '+').replace(/_/g, '/'));
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
        bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
}

// Helper: mathematical modulo (always non-negative)
function mod(a, m) {
    const res = a % m;
    return res >= 0n ? res : res + m;
}

// Modular inverse using extended Euclidean algorithm
function invMod(k, p) {
    if (k === 0n) {
        throw new Error("division by zero");
    }
    if (k < 0n) {
        return p - invMod(-k, p);
    }

    let s = 0n, oldS = 1n;
    let t = 1n, oldT = 0n;
    let r = p, oldR = k;

    while (r !== 0n) {
        const q = oldR / r;
        [oldR, r] = [r, oldR - q * r];
        [oldS, s] = [s, oldS - q * s];
        [oldT, t] = [t, oldT - q * t];
    }

    if (oldR !== 1n) {
        throw new Error("inverse does not exist");
    }

    return mod(oldS, p);
}

// Point doubling: y^2 = x^3 + A*x + B over field P
// point is either INF or [x, y]
function pointDouble(point) {
    if (point === INF) {
        return INF;
    }

    const [x1, y1] = point;
    if (mod(y1, P) === 0n) {
        return INF;
    }

    const m = mod((3n * x1 * x1 + A) * invMod(2n * y1, P), P);
    const x3 = mod(m * m - 2n * x1, P);
    const y3 = mod(m * (x1 - x3) - y1, P);
    return [x3, y3];
}

// Point addition
function pointAdd(p1, p2) {
    if (p1 === INF) {
        return p2;
    }
    if (p2 === INF) {
        return p1;
    }

    const [x1, y1] = p1;
    const [x2, y2] = p2;

    // P + (-P) = INF
    if (x1 === x2 && mod(y1 + y2, P) === 0n) {
        return INF;
    }

    if (x1 === x2 && y1 === y2) {
        return pointDouble(p1);
    }

    const m = mod((y2 - y1) * invMod(mod(x2 - x1, P), P), P);
    const x3 = mod(m * m - x1 - x2, P);
    const y3 = mod(m * (x1 - x3) - y1, P);
    return [x3, y3];
}

// Scalar multiplication using double-and-add
function scalarMult(k, point) {
    if (point === INF) {
        return INF;
    }

    k = mod(k, N);
    if (k === 0n) {
        return INF;
    }

    let result = INF;
    let addend = point;

    while (k !== 0n) {
        if ((k & 1n) === 1n) {
            result = pointAdd(result, addend);
        }
        addend = pointDouble(addend);
        k >>= 1n;
    }

    return result;
}

function isOnCurve(point) {
    if (point === INF) {
        return true;
    }
    const [x, y] = point;
    return mod(y * y - (x * x * x + A * x + B), P) === 0n;
}

// Stub: you will implement this.
// Must return an object { d: BigInt, x: BigInt, y: BigInt }
async function generatePrivateKey() {

    // generate new ECDSA SECP-p-256 key pair
    let signKey = await crypto.subtle.generateKey(
        {
            name: "ECDSA",
            namedCurve: "P-256",
        },
        true,
        ["sign", "verify"]
    );
    let obj = await crypto.subtle.exportKey("jwk", signKey.privateKey);
    return {
        d: BigInt('0x' + bytesToHex(base64ToBytes(obj.d))),
        x: BigInt('0x' + bytesToHex(base64ToBytes(obj.x))),
        y: BigInt('0x' + bytesToHex(base64ToBytes(obj.y))),
        prv: obj,
        pub: await crypto.subtle.exportKey("jwk", signKey.publicKey),
    };
}

// === Main test logic ===
(async function main() {
    if (!isOnCurve(G)) {
        throw new Error("Base point G is not on curve");
    }

    const { d, x, y, prv, pub } = await generatePrivateKey();

    console.log("d =", d.toString(10));
    console.log("x (given) =", x.toString(10));
    console.log("y (given) =", y.toString(10));

    const calcD = d;// + 1n;

    const [calcX, calcY] = scalarMult(calcD, G);

    console.log("x (computed) =", calcX.toString(10));
    console.log("y (computed) =", calcY.toString(10));
    console.log("x matches:", calcX === x);
    console.log("y matches:", calcY === y);

    // console.log(prv);
    // console.log(pub);

    let prvA = await crypto.subtle.importKey("jwk", {
        key_ops: ['sign'],
        ext: true,
        kty: 'EC',
        x: bytesToBase64(hexToBytes(x.toString(16).padStart(64, '0'))),
        y: bytesToBase64(hexToBytes(y.toString(16).padStart(64, '0'))),
        crv: 'P-256',
        d: bytesToBase64(hexToBytes(d.toString(16).padStart(64, '0')))
    }, {
        name: "ECDSA",
        namedCurve: "P-256"
    }, true, ['sign']);

    let pubA = await crypto.subtle.importKey("jwk", {
        key_ops: ['verify'],
        ext: true,
        kty: 'EC',
        x: bytesToBase64(hexToBytes(x.toString(16).padStart(64, '0'))),
        y: bytesToBase64(hexToBytes(y.toString(16).padStart(64, '0'))),
        crv: 'P-256'
    }, {
        name: 'ECDSA',
        namedCurve: 'P-256'
    }, true, ['verify']);

    let prvB = await crypto.subtle.importKey("jwk", {
        key_ops: ['sign'],
        ext: true,
        kty: 'EC',
        x: bytesToBase64(hexToBytes(calcX.toString(16).padStart(64, '0'))),
        y: bytesToBase64(hexToBytes(calcY.toString(16).padStart(64, '0'))),
        crv: 'P-256',
        d: bytesToBase64(hexToBytes(calcD.toString(16).padStart(64, '0')))
    }, {
        name: "ECDSA",
        namedCurve: "P-256"
    }, true, ['sign']);

    let pubB = await crypto.subtle.importKey("jwk", {
        key_ops: ['verify'],
        ext: true,
        kty: 'EC',
        x: bytesToBase64(hexToBytes(calcX.toString(16).padStart(64, '0'))),
        y: bytesToBase64(hexToBytes(calcY.toString(16).padStart(64, '0'))),
        crv: 'P-256'
    }, {
        name: 'ECDSA',
        namedCurve: 'P-256'
    }, true, ['verify']);

    let sigA = await crypto.subtle.sign({
        name: 'ECDSA',
        hash: { name: 'SHA-256' }
    }, prvA, new TextEncoder().encode("Hello World"));

    let sigB = await crypto.subtle.sign({
        name: 'ECDSA',
        hash: { name: 'SHA-256' }
    }, prvB, new TextEncoder().encode("Hello World"));

    console.log("Signature A:", bytesToHex(new Uint8Array(sigA)));
    console.log("Signature B:", bytesToHex(new Uint8Array(sigB)));

    for (let key of [pubA, pubB]) {
        for (let sig of [sigA, sigB]) {
            let valid = await crypto.subtle.verify({
                name: 'ECDSA',
                hash: { name: 'SHA-256' }
            }, key, sig, new TextEncoder().encode("Hello World"));
            console.log("Signature valid:", valid);
        }
    }

})();
